### PR TITLE
Add maximum reccomended version #tgstation/TerraGov-Marine-Corps/pull/5989

### PR DESCRIPTION
**Pull request by ghost**
## About The Pull Request

Ports tgstation/TerraGov-Marine-Corps#5989

title

## Changelog
:cl: TiviPlus, BeloneX
server: Add maximum recommended version of byond
/:cl:

### CODE REVIEW

File: code/_compile_options.dm [hunk 1]
```diff
@@ -50,10 +50,18 @@
...
 #if DM_VERSION < MIN_COMPILER_VERSION || DM_BUILD < MIN_COMPILER_BUILD
 //Don't forget to update this part
-#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update. 
+#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update.
 #error You need version 513.1514 or higher.
 #endif
 
+//Update this whenever the byond version is stable so people stop updating to hilariously broken versions
+#define MAX_COMPILER_VERSION 513
+#define MAX_COMPILER_BUILD 1542
```
Comment (park66665) [inline code/_compile_options.dm]:
```suggestion
#define MAX_COMPILER_BUILD 1536
```

Our recommended host version is 1536. Anything higher than that is very likely to break.

File: code/modules/client/client_procs.dm [hunk 2]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			else
 				qdel(src)
 				return
-
+	if(byond_build > MAX_RECOMMENDED_CLIENT)
+		to_chat(src, "<span class='userdanger'>Your version of byond is likely to be very buggy.</span>")
+		to_chat(src, "<span class='danger'>Please download a old version of byond. If [byond_build] is the latest, you can go to <a href=\"https://secure.byond.com/download/build\">BYOND's website</a> to download other versions.</span>")
```
Comment (park66665) [inline code/modules/client/client_procs.dm]: No info on maximum recommended client build version?

File: code/modules/client/client_procs.dm [hunk 3]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			else
 				qdel(src)
 				return
-
+	if(byond_build > MAX_RECOMMENDED_CLIENT)
+		to_chat(src, "<span class='userdanger'>Your version of byond is over the recommended version MAX_COMPILER_VERSION:MAX_COMPILER_BUILD and is likely to be very buggy.</span>")
+		to_chat(src, "<span class='danger'>Please download a old version of byond. If [byond_build] is the latest, you can go to <a href=\"https://secure.byond.com/download/build\">BYOND's website</a> to download other versions.</span>")
```
Comment (park66665) [inline code/modules/client/client_procs.dm]:
```suggestion
		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended build number for clients([MAX_RECOMMENDED_CLIENT]) and is likely to be very buggy.</span>")
		to_chat(src, "<span class='danger'>Please download an older version of byond. If [byond_build] is the latest, you can go to <a href=\"https://secure.byond.com/download/build\">BYOND's website</a> to download other versions.</span>")
```

File: code/_compile_options.dm [hunk 4]
```diff
@@ -50,10 +50,18 @@
...
-#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update. 
+#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update.
 #error You need version 513.1514 or higher.
 #endif
 
+//Update this whenever the byond version is stable so people stop updating to hilariously broken versions
+#define MAX_COMPILER_VERSION 513
+#define MAX_COMPILER_BUILD 1536
+#if DM_VERSION > MAX_COMPILER_VERSION || DM_BUILD > MAX_COMPILER_BUILD
+#warn WARNING! your byond version is over the recommended version(MAX_COMPILER_VERSION:MAX_COMPILER_BUILD)! There may be unexpected byond bugs!
```
Comment (park66665) [inline code/_compile_options.dm]:
```suggestion
#warn WARNING! your BYOND version is over the recommended version(513.1536)! There may be unexpected byond bugs!
```

```suggestion
#warn WARNING! your BYOND version is over the recommended version! There may be unexpected byond bugs!
```

We can't use defines here; we have to update it manually or just not write it.

File: code/_compile_options.dm [hunk 5]
```diff
@@ -50,10 +50,18 @@
...
-#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update. 
+#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update.
 #error You need version 513.1514 or higher.
 #endif
 
+//Update this whenever the byond version is stable so people stop updating to hilariously broken versions
+#define MAX_COMPILER_VERSION 513
+#define MAX_COMPILER_BUILD 1536
+#if DM_VERSION > MAX_COMPILER_VERSION || DM_BUILD > MAX_COMPILER_BUILD
+#warn WARNING! your BYOND version is over the recommended version(513.1536)! There may be unexpected byond bugs!
```
Comment (park66665) [inline code/_compile_options.dm]:
```suggestion
#warn WARNING! your BYOND version is over the recommended version(513.1536)! There may be unexpected bugs!
```

**Review by park66665 — DISMISSED** [review]

File: code/_compile_options.dm [hunk 6]
```diff
@@ -50,10 +50,18 @@
...
-#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update. 
+#error Your version of BYOND is too out-of-date to compile this project. Go to https://secure.byond.com/download and update.
 #error You need version 513.1514 or higher.
 #endif
 
+//Update this whenever the byond version is stable so people stop updating to hilariously broken versions
+#define MAX_COMPILER_VERSION 513
+#define MAX_COMPILER_BUILD 1536
+#if DM_VERSION > MAX_COMPILER_VERSION || DM_BUILD > MAX_COMPILER_BUILD
+#warn WARNING! your BYOND version is over the recommended version(513.1536)! There may be unexpected bugs!
```
Comment (ike709) [inline code/_compile_options.dm]:
```suggestion
#warn WARNING: Your BYOND version is over the recommended version (513.1536)! Stability is not guaranteed.
```

File: code/modules/client/client_procs.dm [hunk 7]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			else
 				qdel(src)
 				return
-
+	if(byond_build > MAX_RECOMMENDED_CLIENT)
+		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended build number for clients([MAX_RECOMMENDED_CLIENT]) and is likely to be very buggy.</span>")
```
Comment (ike709) [inline code/modules/client/client_procs.dm]:
```suggestion
		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended build number for clients ([MAX_RECOMMENDED_CLIENT]) and is likely to be very buggy.</span>")
```

File: code/modules/client/client_procs.dm [hunk 8]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
...
+		to_chat(src, "<span class='danger'>Please download an older version of byond. If [byond_build] is the latest, you can go to <a href=\"https://secure.byond.com/download/build\">BYOND's website</a> to download other versions.</span>")
```
Comment (ike709) [inline code/modules/client/client_procs.dm]:
```suggestion
		to_chat(src, "<span class='danger'>Please download an older version of byond. You can go to <a href=\"https://secure.byond.com/download/build\">BYOND's website</a> to download other versions.</span>")
```

File: code/modules/client/client_procs.dm [hunk 9]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			else
 				qdel(src)
 				return
-
+	if(byond_build > MAX_RECOMMENDED_CLIENT)
+		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended build number for clients ([MAX_RECOMMENDED_CLIENT]) and is likely to be very buggy.</span>")
```
Comment (utilbird) [inline code/modules/client/client_procs.dm]:
```suggestion
		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended build version for clients ([MAX_RECOMMENDED_CLIENT]) and may be unstable.</span>")
```

Comment (park66665) [reply code/modules/client/client_procs.dm]: "build number" is used more often and sounds more natural to me.

File: code/modules/client/client_procs.dm [hunk 10]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			else
 				qdel(src)
 				return
-
+	if(byond_build > MAX_RECOMMENDED_CLIENT)
+		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended build version for clients ([MAX_RECOMMENDED_CLIENT]) and may be unstable.</span>")
```
Comment (park66665) [inline code/modules/client/client_procs.dm]:
```suggestion
		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended version for clients ([MAX_RECOMMENDED_CLIENT]) and may be unstable.</span>")
```

File: code/modules/client/client_procs.dm [hunk 11]
```diff
@@ -312,7 +312,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 			else
 				qdel(src)
 				return
-
+	if(byond_build > MAX_RECOMMENDED_CLIENT)
+		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended version for clients ([MAX_RECOMMENDED_CLIENT]) and may be unstable.</span>")
```
Comment (park66665) [inline code/modules/client/client_procs.dm]:
```suggestion
		to_chat(src, "<span class='userdanger'>Your version of byond is over the maximum recommended version for clients (build [MAX_RECOMMENDED_CLIENT]) and may be unstable.</span>")
```

Sorry for that. This PR will be merged in short order.

**Review by park66665 — APPROVED** [review]
#3767 is taking time, so is the stabilization of BYOND 514.

All clear. This PR is good to go.

**Pull request merged**